### PR TITLE
allow for arbitrary array types in CallbackCache

### PR DESCRIPTION
**Pull request by ChrisRackauckas**

### CODE REVIEW

File: src/callbacks.jl [hunk 1]
```diff
@@ -637,7 +637,15 @@ mutable struct CallbackCache{conditionType,signType}
   prev_sign::signType
 end
 
-function CallbackCache(max_len,conditionType::Type,signType::Type)
+function CallbackCache(u,max_len,::Type{conditionType},::Type{signType}) where {conditionType,signType}
```
Comment (devmotion) [inline src/callbacks.jl]: What's the purpose of the `where` clauses? IMO all type annotations could/should be removed.

File: src/callbacks.jl [hunk 2]
```diff
@@ -637,7 +637,15 @@ mutable struct CallbackCache{conditionType,signType}
...
+    tmp_condition = similar(u, conditionType, max_len)
+    previous_condition = similar(u, conditionType, max_len)
+    next_sign = similar(u, signType, max_len)
+    prev_sign = similar(u, signType, max_len)
+    CallbackCache{typeof(tmp_condition),typeof(next_sign)}(tmp_condition,previous_condition,next_sign,prev_sign)
+end
+
+function CallbackCache(max_len,::Type{conditionType},::Type{signType}) where {conditionType,signType}
```
Comment (devmotion) [inline src/callbacks.jl]: Same here

File: src/callbacks.jl (hunk 1 above)
Comment (ChrisRackauckas) [reply src/callbacks.jl]: `::Type` doesn't specialize.

Comment (devmotion) [reply src/callbacks.jl]: But isn't it sufficient to remove the type annotation completely?

Comment (ChrisRackauckas) [reply src/callbacks.jl]: IIRC no, you need to specialize it so it transforms it from `DataType`.

Comment (ChrisRackauckas) [reply src/callbacks.jl]: IIRC, this and how `::Function` works are the two exceptions to the auto-specialization rule.

Comment (devmotion) [reply src/callbacks.jl]: Oh interesting, I didn't know that

Comment (devmotion) [reply src/callbacks.jl]: But wait, isn't it slightly different for `Function` in that if the type annotation is removed, it is actually specialized? E.g., `f(g) = g(0)` should specialize whereas `f(g::Function) = g(0)` doesn't?

Comment (ChrisRackauckas) [reply src/callbacks.jl]: Yes, it's the opposite way with `Function`, with the caveat that `Function` will also specialize there if it's called... two oddly specific and weird heuristics.

**Pull request merged**